textsemantics/specific_words.py

```diff
@@ -0,0 +1,265 @@
+    doc2word = list()
+    word2doc = dict()
+
+    for i, tokens in enumerate(tokens_list):
+        sent = Sentence(" ".join(tokens))
+        embedder.embed(sent)
+        doc_emb = np.zeros(embedder.embedding_length)
+        doc2word.append(set())
+        for token in sent.tokens:
+            if token.text not in word2doc:
+                word2doc[token.text] = set()
+            word2doc[token.text].add(i)
+            doc2word[i].add(token.text)
+
+            emb = token.embedding.cpu().detach().numpy()
+            doc_emb += emb / len(tokens)
+            if token.text not in word_embs:
+                word_embs[token.text] = emb
+        doc_embs.append(doc_emb)
+
+    doc_embs = np.array(doc_embs)
+
+    return doc_embs, word_embs, word2doc, doc2word
+
+
+def find_corpus_words(doc_embs: np.ndarray, word_embs: Dict[str, np.ndarray]) -> List[List[Tuple[str, float]]]:
+    # compute distances
+    distances = np.zeros((len(word_embs), doc_embs.shape[0]))
+    words = list(word_embs.keys())
+    for i, word in enumerate(words):
+        for j in range(doc_embs.shape[0]):
+            distances[i, j] = 1 - cos_sim(word_embs[word], doc_embs[j, :])
+
+    # compute scores
+    doc_desc = []
+    for j in range(doc_embs.shape[0]):
+        scores = np.zeros(len(word_embs))
+        for i in range(len(words)):
+            mask = np.full(doc_embs.shape[0], fill_value=True)
+            mask[j] = False
+            scores[i] = distances[i, j] - np.mean(distances[i, mask])
+
+        idx = np.argsort(scores)
+        doc_desc.append([(words[w], scores[w]) for w in idx])
+    return doc_desc
+
+
+def find_document_words(doc_embs: np.ndarray, word_embs: Dict[str, np.ndarray], word2doc: Dict[str, Set[int]], doc2word: List[Set[str]]) -> List[List[Tuple[str, float]]]:
+    distances = dict()
+    for word, w_emb in word_embs.items():
+        for j in word2doc[word]:
+            distances[word, j] = 1 - cos_sim(w_emb, doc_embs[j, :])
+
+    doc_desc = list()
+    for j in range(doc_embs.shape[0]):
+        scores = np.zeros(len(doc2word[j]))
+        ind2word = dict(zip(list(range(len(doc2word[j]))), list(doc2word[j])))
+        for k, word in enumerate(doc2word[j]):
+            # current document and a single word
+            sum_of_distances = sum([distances[word, x] for x in word2doc[word]])
+            if len(word2doc[word]) > 1:
+                mean_distance = (sum_of_distances - distances[word, j]) / (
+                    len(word2doc[word]) - 1
+                )
+            else:
+                mean_distance = 0
+            scores[k] = distances[word, j] - mean_distance
+        idx = np.argsort(scores)
+        doc_desc.append([(ind2word[x], scores[x]) for x in idx])
+    return doc_desc
+
+
+def embedding_corpus_keywords(
+    texts: Optional[List[str]] = None,
+    tokens: Optional[List[List[str]]] = None,
+    language: str = "slovenian",
+):
+    assert bool(texts) != bool(tokens), "Please provide either texts or tokens"
```

Comment:
Če predpostavljamo, da je to knjižnica za slovenščino (hard-codani parametri jezika), bi morda imeli opozorila v slovenščini? Zgolj ideja.

Reply:
Ja je kar smislno glede na to, da so tudi notebooki v slovenščini. Sem spremenil